src/ILCompiler.ReadyToRun/src/Compiler/ReadyToRunTableManager.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+
+using ILCompiler.DependencyAnalysis;
+using ILCompiler.DependencyAnalysis.ReadyToRun;
+using ILCompiler.DependencyAnalysisFramework;
+
+using Internal.TypeSystem;
+
+using Debug = System.Diagnostics.Debug;
+
+namespace ILCompiler
+{
+    public class ReadyToRunTableManager : MetadataManager
+    {
+        public ReadyToRunTableManager(CompilerTypeSystemContext typeSystemContext)
+            : base(typeSystemContext, new NoMetadataBlockingPolicy(), new NoManifestResourceBlockingPolicy()) {}
+
+        public override void AddToReadyToRunHeader(ReadyToRunHeaderNode header, NodeFactory nodeFactory, ExternalReferencesTableNode commonFixupsTableNode)
+        {
+            // We don't attach any metadata blobs.
+        }
+
+        protected override void Graph_NewMarkedNode(DependencyNodeCore<NodeFactory> obj)
+        {
+            base.Graph_NewMarkedNode(obj);
+            
+            var eetypeNode = obj as AvailableType;
+            if (eetypeNode != null)
+            {
+                _typesWithEETypesGenerated.Add(eetypeNode.Type);
+                return;
+            }
+        }
+
+        public override MethodDesc GetCanonicalReflectionInvokeStub(MethodDesc method) => throw new NotImplementedException();
```

Comment:
Does this mean that we'll gradually need to fill in all of these or are they irrelevant for R2R? At the first glance there seems to be some overlap with my current efforts regarding the reverse token lookup vs. upfront registration of external ref tokens during resolution in JIT.

Reply:
To be honest I'm not sure if we'll want to use the stock `MetadataManager` long term or not since I don't think we'll use most of these virtuals. Perhaps we can refactor so that the base manager class just tracks the presence of interesting type system elements and then it can be specialized into being a ECMA metadata manager or a R2R supporting table manager.

Reply:
Sounds reasonable to me, that matches my feeling about NodeFactory. I just think that, if we're supposed to do this properly, R2R needs to become a first-class citizen in the CoreRT source code, not just an appendix. I'm interested in Michal's thoughts in this space. My personal gut feeling is that the partial CorInfoImpl split worked fine and we might be able to set further order to the code by forking NodeFactory and possibly MetadataManager in a similar manner.